### PR TITLE
Custom Parameter Validation within Component

### PR DESCRIPTION
**Pull request by Jbsco**
This pull request adds the optionally implementable `Validate_Parameters` which allows a user to leverage custom inter-parameter validation in addition to individual component validation. This is optional, but components which have parameters will still have the default implementation of the action added to their respective generated templates.

### CODE REVIEW

File: doc/example_architecture/parameter_component/component-parameter_component-implementation.adb [hunk 1]
```diff
@@ -57,4 +57,21 @@ package body Component.Parameter_Component.Implementation is
...
+   -- subprogram in the implementation package is a function that returns "Valid". However, this function can, and should be
+   -- overridden if something special needs to happen to further validate a parameter. Examples of this might be validation of
+   -- certain parameters beyond individual type ranges, or performing other special functionality that only needs to be
+   -- performed after parameters have been validated.
+   overriding function Validate_Parameters_Action (
+      Self : in out Instance;
+      Start_Count : in Packed_U16.U;
+      Hello_World_Value : in Packed_U16.U
+   ) return Parameter_Validation_Status.E is
+      Ignore : Instance renames Self;
```
Comment (dinkelk) [inline doc/example_architecture/parameter_component/component-parameter_component-implementation.adb:70]: You probably also need to ignore the other values here too, no? You can do this by naming them `Ignore_Start_Count`, etc.

File: doc/example_architecture/parameter_component/component-parameter_component-implementation.adb [hunk 2]
```diff
@@ -57,4 +57,21 @@ package body Component.Parameter_Component.Implementation is
...
+      use Parameter_Validation_Status;
+   begin
+      -- TODO: Perform action to validate parameters.
+      return Valid;
+   end Validate_Parameters_Action;
```
Comment (dinkelk) [inline doc/example_architecture/parameter_component/component-parameter_component-implementation.adb:75]: See if we can't just implement this in the .ads.

File: gen/templates/component/component-name.adb [hunk 3]
```diff
@@ -943,12 +943,31 @@ package body Component.{{ name }} is
...
+               {{ par.name }} => Self.Staged_Parameters.Get_{{ par.name }}{{ "," if not loop.last }}
+{% endfor %}
+            ) /= Valid
+            then
+               -- The default return for this function always returns "Valid", so this
+               -- should be inaccessible.
+               Status := Validation_Error;
+            end if;
+            -- Set status:
+            Status := Validation_Success;
```
Comment (dinkelk) [inline gen/templates/component/component-name.adb:970]: Don't you need to put this in an `else`?

File: src/types/parameter/parameter_enums.enums.yaml [hunk 4]
```diff
@@ -55,3 +61,12 @@ enums:
...
         description: The operation could not be performed because it was dropped from a full queue.
+  - name: Parameter_Validation_Status
+    description: This status enumeration provides information on the result of a parameter validation.
+    literals:
+      - name: Valid
+        value: 0
+        description: Parameter was validated.
+      - name: Invalid
+        value: 1
+        description: Parameter was found to be invalid.
```
Comment (dinkelk) [inline src/types/parameter/parameter_enums.enums.yaml:72]: Doesn't this more like validate a set of parameters together. The descriptions sound like this is validating just a single parameter. Maybe reword so it sounds like we are validating a parameter set.

**Review by dinkelk — CHANGES_REQUESTED** [review]

File: doc/example_architecture/parameter_component/component-parameter_component-implementation.adb (hunk 2 above)
Comment (Jbsco) [reply doc/example_architecture/parameter_component/component-parameter_component-implementation.adb]: Moved to `.ads` in [2008566](https://github.com/lasp/adamant/pull/64/commits/2008566276f5ade50bfb38842dcb8107d3c220ff). Tested template, style, and unit tests per affected component.

File: doc/example_architecture/parameter_component/component-parameter_component-implementation.adb (hunk 1 above)
Comment (Jbsco) [reply doc/example_architecture/parameter_component/component-parameter_component-implementation.adb]: Addressed in [2008566](https://github.com/lasp/adamant/pull/64/commits/2008566276f5ade50bfb38842dcb8107d3c220ff). Implementation moved to `.ads`.

File: src/types/parameter/parameter_enums.enums.yaml (hunk 4 above)
Comment (Jbsco) [reply src/types/parameter/parameter_enums.enums.yaml]: Reworded in [87b53e2](https://github.com/lasp/adamant/pull/64/commits/87b53e25918b9ba58654c3b49c2e6f54074784cb).

File: gen/templates/component/component-name.adb (hunk 3 above)
Comment (Jbsco) [reply gen/templates/component/component-name.adb]: Addressed in [87b53e2](https://github.com/lasp/adamant/pull/64/commits/87b53e25918b9ba58654c3b49c2e6f54074784cb).

File: gen/templates/component/component-name.ads [hunk 5]
```diff
@@ -269,6 +269,18 @@ package Component.{{ name }} is
    -- been updated.
    not overriding procedure Update_Parameters_Action (Self : in out Base_Instance) is abstract;
 
+   -- This function is called when the parameter operation type is "Validate". The default implementation of this
+   -- subprogram in the implementation package is a function that returns "Valid". However, this function can, and should be
+   -- overridden if something special needs to happen to further validate a parameter. Examples of this might be validation of
+   -- certain parameters beyond individual type ranges, or performing other special functionality that only needs to be
+   -- performed after parameters have been validated.
```
Comment (dinkelk) [inline gen/templates/component/component-name.ads:276]: Add that range checking is part of staging, so doesn't need to be implemented by the user in this function.

File: gen/templates/component/component-name.ads [hunk 6]
```diff
@@ -269,6 +269,18 @@ package Component.{{ name }} is
...
+   not overriding function Validate_Parameters_Action (
```
Comment (dinkelk) [inline gen/templates/component/component-name.ads:277]: Name it `Validate_Parameters` instead.

File: gen/templates/component/component-name.adb [hunk 7]
```diff
@@ -943,12 +943,32 @@ package body Component.{{ name }} is
...
+            -- each parameter individually, this function is used for extended,
+            -- user-implemented validation.
+            if Base_Instance'Class (Self).Validate_Parameters_Action (
+{% for par in parameters %}
+               {{ par.name }} => Self.Staged_Parameters.Get_{{ par.name }}{{ "," if not loop.last }}
+{% endfor %}
+            ) /= Valid
+            then
+               -- The default return for this function always returns "Valid", so this
+               -- should be inaccessible.
```
Comment (dinkelk) [inline gen/templates/component/component-name.adb:966]: Remove this comment

**Review by dinkelk — CHANGES_REQUESTED** [review]

File: gen/templates/component/component-name.ads (hunk 5 above)
Comment (Jbsco) [reply gen/templates/component/component-name.ads]: Added in [c61b1f5](https://github.com/lasp/adamant/pull/64/commits/c61b1f53a99257b70b15bbd0a22495ca6cccd935).

File: gen/templates/component/component-name.adb (hunk 7 above)
Comment (Jbsco) [reply gen/templates/component/component-name.adb]: Addressed in [c61b1f5](https://github.com/lasp/adamant/pull/64/commits/c61b1f53a99257b70b15bbd0a22495ca6cccd935).

File: gen/templates/component/component-name.ads (hunk 6 above)
Comment (Jbsco) [reply gen/templates/component/component-name.ads]: Addressed in [c61b1f5](https://github.com/lasp/adamant/pull/64/commits/c61b1f53a99257b70b15bbd0a22495ca6cccd935).

File: src/types/parameter/parameter_enums.enums.yaml [hunk 8]
```diff
@@ -28,6 +31,9 @@ enums:
       - name: Length_Error
         value: 3
         description: Parameter length was not correct.
+      - name: Validation_Success
+        value: 4
+        description: Parameter values were successfully validated.
```
Comment (dinkelk) [inline src/types/parameter/parameter_enums.enums.yaml:36]: Typically we only include a single enumeration value for "success", and everything else can be assumed to be an off nominal return. Is there an issue with using the Success/0 in this case to indicate successful validation?

**Review by dinkelk — CHANGES_REQUESTED** [review]

File: src/types/parameter/parameter_enums.enums.yaml (hunk 8 above)
Comment (Jbsco) [reply src/types/parameter/parameter_enums.enums.yaml]: No issue, it makes sense to remove this. Addressed in [9328969](https://github.com/lasp/adamant/pull/64/commits/9328969af2f37b9ecf0b0c96d383d93f2ecd69b8).

File: gen/templates/component/component-name.adb [hunk 9]
```diff
@@ -943,12 +943,31 @@ package body Component.{{ name }} is
...
+               {{ par.name }} => Self.Staged_Parameters.Get_{{ par.name }}{{ "," if not loop.last }}
+{% endfor %}
+            ) /= Valid
+            then
+               -- Set status:
+               Status := Validation_Error;
+            else
+               -- Set status:
+               Status := Success;
+            end if;
```
Comment (dinkelk) [inline gen/templates/component/component-name.adb:970]: This would be much clearer as a case statement, since it will show all the possible returns.

**Review by dinkelk — CHANGES_REQUESTED** [review]

File: gen/templates/component/component-name.adb (hunk 9 above)
Comment (Jbsco) [reply gen/templates/component/component-name.adb]: Agreed! Addressed in [1b827a7](https://github.com/lasp/adamant/pull/64/commits/1b827a780e24f58930e9c35e67bc685087cccec5).

**Review by dinkelk — APPROVED** [review]
